c/parallel/parallel_for_ordered.cc

```diff
@@ -0,0 +1,298 @@
+//------------------------------------------------------------------------------
+// Copyright 2019 H2O.ai
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//------------------------------------------------------------------------------
+#include <mutex>
+#include <thread>   // std::this_thread
+#include <vector>
+#include "parallel/api.h"
+#include "parallel/thread_scheduler.h"
+#include "parallel/thread_team.h"
+#include "utils/assert.h"
+#include "utils/exceptions.h"
+#include "utils/function.h"
+namespace dt {
+
+
+
+//------------------------------------------------------------------------------
+// ordered_task
+//------------------------------------------------------------------------------
+using f1t = function<void(size_t)>;
+static f1t noop = [](size_t) {};
+
+class ordered_task : public thread_task {
+  private:
+    static constexpr size_t READY_TO_START = 0;
+    static constexpr size_t STARTED = 1;
+    static constexpr size_t READY_TO_ORDER = 2;
+    static constexpr size_t READY_TO_FINISH = 3;
```

Comment:
(a) it's internal, so doesn't really matter;
(b) I need to do computations on `state` in `advance_state()`, so it's more convenient to keep it as an integer;
(c) enum would make the expressions longer, because then you'd have to prefix each constant with the name of the enum.

Reply:
Yes, I noticed you do (b), so it seems like it is easy to mess it up accidentally...